src/main/kotlin/com/terraformation/backend/tracking/db/PlantingSiteImporter.kt

```diff
@@ -386,83 +376,167 @@ class PlantingSiteImporter(
-              .mapIndexed { index, polygon ->
-                ShapefileFeature(
-                    polygon,
-                    mapOf(
-                        PLOT_NAME_PROPERTY to "${index + 1}",
-                        SUBZONE_NAME_PROPERTY to subzoneFeature.properties[SUBZONE_NAME_PROPERTY]!!,
-                        ZONE_NAME_PROPERTY to subzoneFeature.properties[ZONE_NAME_PROPERTY]!!,
-                    ),
-                    crs)
+  /**
+   * Assigns monitoring plots to subzones and to clusters of permanent monitoring plots.
+   *
+   * If all 4 plots in a cluster fall in the same subzone, they are eligible to be used as permanent
+   * monitoring plots.
+   *
+   * The permanent monitoring plots for a planting zone must always come in clusters of 4, and those
+   * clusters of 4 must be selected at random. The number of permanent monitoring plots can vary
+   * from one observation to the next, but the plots that are selected have to be consistent over
+   * time.
+   *
+   * We guarantee this consistency by randomizing the list of eligible clusters at import time and
+   * recording the randomized order in the form of a cluster number on each monitoring plot. The
+   * cluster number starts at 1 for each planting zone, and there are always exactly 4 plots with a
+   * given cluster number in a given planting zone.
+   *
+   * The cluster numbers allow permanent monitoring plots to be allocated using a SQL query. If we
+   * want 5 clusters of permanent plots from planting zone 4321, we can do a query along the lines
+   * of
+   *
+   *     SELECT mp.id
+   *     FROM tracking.monitoring_plots mp
+   *     JOIN tracking.planting_subzones ps ON mp.planting_subzone_id = ps.id
+   *     WHERE ps.planting_zone_id = 4321
+   *     AND mp.permanent_cluster <= 5;
+   *
+   * and get back a list of all 20 monitoring plots in 5 randomly-chosen 2x2 clusters.
+   */
+  private fun assignPlots(
+      allPlots: Collection<Collection<Polygon>>,
+      subzoneRows: Collection<PlantingSubzonesRow>,
+  ): List<MonitoringPlotsRow> {
+    val now = clock.instant()
+    val userId = currentUser().userId
+
+    if (subzoneRows.map { it.plantingZoneId }.distinct().size != 1) {
+      throw IllegalArgumentException("BUG! Plots must be assigned one zone at a time.")
+    }
+
+    // Eliminate any monitoring plots that straddle subzone boundaries, and record which subzone
+    // each plot is in as well as its plot number within that subzone. Plots are still grouped in
+    // 2x2 clusters at this point (though a given cluster might have fewer than 4 plots if some of
+    // them didn't fall within the subzone).
+    val clustersInSubzones: List<List<MonitoringPlotsRow>> =
+        subzoneRows.flatMap { subzoneRow ->
+          var plotNumber = 0
+
+          val subzoneClusters =
+              allPlots
+                  .map { cluster -> cluster.filter { it.coveredBy(subzoneRow.boundary!!) } }
```

Comment:
do we want to remove clusters that were already considered for a subzone row - from allPlots? (for runtime perf, only if it makes sense. . i understand there's a cost to the list manipulation... and these are one-time imports.)

Reply:
That's definitely a reasonable question! It would make the data flow a bit more complex, but I'll try it locally and see how it looks.

Reply:
Thanks for giving this a try. Makes sense that the db operations take most of the time.